wade-review.md

```diff
@@ -0,0 +1,596 @@
+
+        return builder.build(properties);
+    }
+
+
+}
+```
+
+ I’ll chain on another peek method to log if our data is processed successfully, then we’ll send it to the output topic with the .to method.  
+
+ 
+```java
+    public Topology topology(Properties properties) {
+
+        final ValueMapper<JsonNode, GitHubPRInfo> valueMapper = jsonNode -> {
+
+            return new GitHubPRInfo(state, createdAt, closedAt, prNumber);
+        };
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        builder.stream(INPUT_TOPIC, Consumed.with(Serdes.String(), jsonNodeSerde))
+                .peek((key, value) -> LOG.info("Incoming key[{}] value[{}]", key, value))
+                .mapValues(valueMapper)
+                .process(new MyProcessorSupplier())
+                .peek((key, value) -> LOG.info("Outgoing value key[{}] value[{}]", key, value))
+                .to(OUTPUT_TOPIC, Produced.with(Serdes.String(), prStateCounterSerde));
+
+        return builder.build(properties);
+    }
+
+
+}
+```
+
+Now, I want to make the data coming in a more convenient shape, and I’ll use a ValueMapper. It’ll take the data coming in from the connector and give it a more convenient shape. I’ll use the GitHubPRInfo model for doing that.  
+
+
+```java
+    public Topology topology(Properties properties) {
+
+        final ValueMapper<JsonNode, GitHubPRInfo> valueMapper = jsonNode -> {
+            JsonNode prStateVal = jsonNode.findValue("state");
+            JsonNode prCreatedAtVal = jsonNode.findValue("created_at");
+            JsonNode prClosedAtVal = jsonNode.findValue("closed_at");
+            JsonNode prNumberVal = jsonNode.findValue("number");
+            String state = prStateVal != null ? prStateVal.asText() : NO_STATE;
+            String createdAt = prCreatedAtVal != null ? prCreatedAtVal.asText() : "no-created-at";
+            String closedAt = prClosedAtVal != null ? prClosedAtVal.asText() : "no-closed-at";
+            int prNumber = prNumberVal != null ? prNumberVal.asInt() : Integer.MIN_VALUE;
+
+            return new GitHubPRInfo(state, createdAt, closedAt, prNumber);
+        };
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        builder.stream(INPUT_TOPIC, Consumed.with(Serdes.String(), jsonNodeSerde))
+                .peek((key, value) -> LOG.info("Incoming key[{}] value[{}]", key, value))
+                .mapValues(valueMapper)
+                .process(new MyProcessorSupplier())
+                .peek((key, value) -> LOG.info("Outgoing value key[{}] value[{}]", key, value))
+                .to(OUTPUT_TOPIC, Produced.with(Serdes.String(), prStateCounterSerde));
+
+        return builder.build(properties);
+    }
+
+
+}
+```
+
+The 4 pieces of information we’re taking in as JsonNodes are the PR’s state, when it was created, and when it was closed, as well as the PR number. We then transform those to strings and ints, making sure to provide a case if the value happens to be null. 
```

Comment:
I mentioned it earlier, but ideally, start off the video with an image of the JSON or something explaining what we get. That way, when we get here, some of this has already been explained at people have a better understanding of the structure we are reading.

Reply:
https://docs.google.com/document/d/1C0U9B_UAVB1NVAa3fuCT2yF606TdHCt8Q5iBjen02Vk/edit?disco=AAABJ2-4-z4